### PR TITLE
feat(vc): Create verifiable credentials

### PR DESCRIPTION
**Pull request by l1h3r**
Adds structures for representing Verifiable Credentials and Presentations according to the [Verifiable Credentials Data Model (1.0)](https://www.w3.org/TR/2019/REC-vc-data-model-20191119).

~Depends on https://github.com/iotaledger/identity.rs/pull/24~ (merged)

### CODE REVIEW

File: identity_vc/src/credential.rs
Comment (l1h3r) [inline identity_vc/src/credential.rs:10]: Several fields here are represented as `Object`. Most of them should probably be structs as they do have a bit of a pre-defined structure.

An example of how `Evidence` could be represented.

```rust
#[derive(Debug, Deserialize, Serialize)]
pub struct Evidence {
  id: Option<String>,
  #[serde(rename = "type")]
  types_: String,
  #[serde(flatten)]
  object: Object,
}
```

Since many fields are optional and no data is being discarded, I don't think it's too high of a priority.

@tensor-programming thoughts?

Comment (tensor-programming) [reply identity_vc/src/credential.rs:13]: Its up to you whether or not you want to include them.  I see you are basically nesting hashmaps for the object representation which is fine.  The one thing I would worry about is allowing convoluted json structures to be valid since this doesn't make any assumptions about the types of keys you are accepting. Aside from that, it should be acceptable.

File: identity_vc/src/common/oom.rs
Comment (tensor-programming) [reply identity_vc/src/common/oom.rs:6]: You could write a custom Serializer/Deserializer to handle the cases where you get only one piece of data as opposed to a larger value.  That's what I did for the Context in the DIDDocument.  If the vector only contains one value, represent it as a single string in the Json and if it contains more than one value, represent it as a list.  In this case, it would be one String vs Multiple nested objects.

File: identity_vc/src/common/context.rs
Comment (tensor-programming) [reply identity_vc/src/common/context.rs:10]: I see.  Still need to see a use case where the context is a full JSON object but am sure that is a fine implementation for that scenario.

**Comment by tensor-programming** [comment]
Will review this tomorrow as soon as I can.

**Review by JelleMillenaar — COMMENTED** [review]
Looking very good to me. Extensive implementation including all the optional fields, so I am very happy with that :) Would be good to wait for a review from @tensor-programming to make sure coding styles are aligned. I would like DID and VC core impl to feel very compatible.

File: identity_vc/src/common/oom.rs
Comment (l1h3r) [reply identity_vc/src/common/oom.rs:6]: Yes, I think I just forgot about it when I moved some others to `identity_core::common`.

A top-level `identity_common` library would decouple these from the DID implementation; I'm all for it.

File: identity_vc/src/common/context.rs
Comment (l1h3r) [reply identity_vc/src/common/context.rs:5]: The DID standard doesn't mention object-based contexts but the implementation **could** be flexible. The DID document could use *this* `Context` (`context: OneOrMany<Context>`) with a few minor changes.

It may not be entirely useful without JSON-LD support **but** avoids ser/deserialization errors on spec-compliant data.

@tensor-programming thoughts?

Comment (l1h3r) [reply identity_vc/src/common/context.rs]: @JelleMillenaar I put this particular change in a [new branch](https://github.com/iotaledger/identity.rs/tree/feat/shared-context). It's a bit more involving with `serde_diff` and depends on diff support for `Value` which @tensor-programming is currently working on(?)

Comment (tensor-programming) [reply identity_vc/src/common/context.rs]: Yeah sorry, didn't see this conversation until just now.  Yeah I am working on the diff implementation and I will do something about making serde's `Value` a priority.  The context as it is looks good right now and we can always add more flexibility later.  Ill do a full review in the morning.

File: identity_vc/src/macros.rs
Comment (tensor-programming) [inline identity_vc/src/macros.rs:1]: I wonder if we should potentially move these macros over to core or maybe build a proc macro for builders in the derive crate.  Just a thought since if we want to follow the builder pattern then it may be a good idiom to adopt across all of the crates. There are also some battle tested builder derive crates already in the rust ecosystem we could look at to avoid building our own. `rust-derive-builder` and `typed_builder` come to mind.  Again, just a consideration, you implementation looks flexible enough.

Comment (l1h3r) [reply identity_vc/src/macros.rs:1]: `rust-derive-builder` looks like a nice option - it supports *almost* everything **but** it doesn't have this special handling for `Vec<T>` (which ultimately becomes `OneOrMany<T>`). `typed_builder` is missing `TryInto` support, which is really nice to have.

I'll move to `identity_core` for now; maybe look at a fork of `rust-derive-builder` if our needs outgrow these macros (it needs a few changes anyways, eg. supporting `try_setter` with `strip_option`).

`#[derive(Builder)]` is definitely a **much** nicer pattern though.

Comment (tensor-programming) [reply identity_vc/src/macros.rs]: Alright, sounds like a good plan.  When you merge I will incorporate the pattern into the core for consistency.

File: identity_vc/src/lib.rs
Comment (l1h3r) [reply identity_vc/src/lib.rs:1]: I guess the change from `#[macro_use] extern crate` to `use` is quite consistent with other items so 👍 .

I think the `mod` change was a disaster 👎  I still don't see that structure often in practice and I think it makes the mental model worse. I thought it was largely added to eliminate the restriction of **requiring** `mod.rs` when using sub-modules.

**Comment by l1h3r** [comment]
@JelleMillenaar started a [new branch](https://github.com/iotaledger/identity.rs/tree/feat/shared-context) to unify the DID/Credential Context types - I think this PR is good for now as a base, wdyt? (tagged you again for review)

@tensor-programming updated to reflect some of your latest recommendations

File: identity_vc/src/lib.rs
Comment (tensor-programming) [reply identity_vc/src/lib.rs]: I actually was really glad when they decided to remove the mod files.  Though admittedly, I am not a huge fan of the Rust module system in general.  That you have to create a file when you want to use a folder whether or it be inside or outside of the folder is always going to feel a little awkward.  Still its a bit less confusing to have distinctly named files rather than 30 files called mod.rs especially if you have only that mod file in the folder.

**Review by tensor-programming — APPROVED** [review]

**Review by JelleMillenaar — APPROVED** [review]

**Pull request merged**
